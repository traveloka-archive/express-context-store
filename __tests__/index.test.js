/* eslint-env jest */
const createMiddleware = require('../');

test('call next inside middleware', () => {
  const req = {};
  const res = {};
  const next = jest.fn();
  const middleware = createMiddleware();
  middleware(req, res, next);

  expect(next).toHaveBeenCalledTimes(1);
});

test('provide Map-like API in req.context', () => {
  const req = {};
  const res = {};
  const next = jest.fn();
  const middleware = createMiddleware();
  middleware(req, res, next);

  expect(req.context.get('key')).toBe(null);
  req.context.set('key', 'value');
  expect(req.context.get('key')).toBe('value');
});

test('throw error if context is written multiple times', () => {
  const req = {};
  const res = {};
  const next = jest.fn();
  const middleware = createMiddleware();
  middleware(req, res, next);

  req.context.set('key', 'value');
  const fn = () => req.context.set('key', 'another');
  expect(fn).toThrow('key already exist in req.context');
});

test('allow key to be written multiple times explicitly', () => {
  const req = {};
  const res = {};
  const next = jest.fn();
  const middleware = createMiddleware();
  middleware(req, res, next);

  req.context.set('key', 'value', { writable: true });
  expect(req.context.get('key')).toBe('value');

  req.context.set('key', 'another', { writable: true });
  expect(req.context.get('key')).toBe('another');

  req.context.set('key', 'different', { writable: true });
  expect(req.context.get('key')).toBe('different');
});

test('require writable option to be provided for multiple write', () => {
  const req = {};
  const res = {};
  const next = jest.fn();
  const middleware = createMiddleware();
  middleware(req, res, next);

  req.context.set('key', 'value', { writable: true });
  req.context.set('key', 'another');
  const fn = () => req.context.set('key', 'different', { writable: true });
  expect(fn).toThrow('key already exist in req.context');
});

test('provides toObject method for serialization', () => {
  const req = {};
  const res = {};
  const next = jest.fn();
  const middleware = createMiddleware();
  middleware(req, res, next);

  req.context.set('a', 'b');
  req.context.set('c', { d: 'e' });
  req.context.set('f', ['g', 'h']);
  req.context.set('i', 1);
  req.context.set('j', true);

  const object = req.context.toObject();
  const expectedObject = {
    a: 'b',
    c: {
      d: 'e',
    },
    f: ['g', 'h'],
    i: 1,
    j: true,
  };
  expect(object).toEqual(expectedObject);
});

test('fallback to Map for unknown getter', () => {
  const req = {};
  const res = {};
  const next = jest.fn();
  const middleware = createMiddleware();
  middleware(req, res, next);

  req.context.set('key', 'value');
  expect(req.context.size).toBe(1);
  expect(req.context.has('key')).toBe(true);

  // conversion to iterable
  expect(Array.from(req.context.keys())).toMatchSnapshot('.keys()');
  expect(Array.from(req.context.values())).toMatchSnapshot('.values()');
  expect(Array.from(req.context.entries())).toMatchSnapshot('.entries()');

  req.context.delete('key');
  expect(req.context.size).toBe(0);
});

test('custom property name', () => {
  const req = {};
  const res = {};
  const next = jest.fn();
  const middleware = createMiddleware({ property: 'data' });
  middleware(req, res, next);

  expect(req.context).not.toBeDefined();

  req.data.set('key', 'value');
  expect(req.data.get('key')).toBe('value');
});
