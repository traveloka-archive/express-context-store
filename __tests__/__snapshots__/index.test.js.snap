// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fallback to Map for unknown getter: .entries() 1`] = `
Array [
  Array [
    "key",
    Object {
      "readOnly": true,
      "value": "value",
    },
  ],
]
`;

exports[`fallback to Map for unknown getter: .keys() 1`] = `
Array [
  "key",
]
`;

exports[`fallback to Map for unknown getter: .values() 1`] = `
Array [
  Object {
    "readOnly": true,
    "value": "value",
  },
]
`;
